SystemServices/SystemServices.cpp

```diff
@@ -2397,10 +2398,12 @@ namespace WPEFramework {
 					if( dirExists(path+country)  && Utils::fileExists(city.c_str()) ) 
 					{
 						if (!dirExists(dir)) {
-							std::string command = "mkdir -p " + dir + " \0";
+							command = "mkdir -p " + dir + " \0";
 							Utils::cRunScript(command.c_str());
 						} else {
 							//Do nothing//
+							command = "cat /usr/share/zoneinfo/" +timeZone +" > /usr/share/zoneinfo/Universal" +" \0";
```

Comment:
Could you clarify why the redirect is set to /usr/share/zoneinfo/Universal?

Reply:
Hi @santoshcomcast 
As per the changes, we are redirecting the timezone to /usr/share/zoneinfo/Universal file. This is done because /usr/share/zoneinfo/Universal is the softlink for device time by default.

`lrwxrwxrwx    1 root     root          29 Jul 31 13:45 localtime -> /usr/share/zoneinfo/Universal`
/etc/localtime is linked to this file and whatever timezone present in /usr/share/zoneinfo/Universal will be set as system timezone.

Reply:
What is time-zone set in that device?

I have tried set to time-zone = America/New_York  then /etc/localtime soft link  to  "/usr/share/zoneinfo/America/New_York" same I have observer with time-zone= Universal.

`root@element-teone:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.setTimeZoneDST", "params":{"timeZone": "America/New_York"}} '
{"jsonrpc":"2.0","id":3,"result":{"success":true}}root@element-teone:~#
root@element-teone:~#
root@element-teone:~#
root@element-teone:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.getTimeZoneDST"} '
{"jsonrpc":"2.0","id":3,"result":{"timeZone":"America\/New_York","accuracy":"FINAL","success":true}}root@element-teone:~#

root@element-teone:~# ls -la /etc/localtime
lrwxrwxrwx    1 root     root            25 Jul  9 16:09 /etc/localtime -> /opt/persistent/localtime
root@element-teone:~# ls -la /opt/persistent/localtime
lrwxrwxrwx    1 root     root            36 Aug  8 06:45 /opt/persistent/localtime -> /usr/share/zoneinfo/America/New_York`


`root@element-teone:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.setTimeZoneDST", "params":{"timeZone": "Universal"}} '
root@element-teone:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.getTimeZoneDST"} '
{"jsonrpc":"2.0","id":3,"result":{"timeZone":"Universal","accuracy":"FINAL","success":true}}root@element-teone:~#
root@element-teone:~#
root@element-teone:~#
root@element-teone:~# ls -la /etc/localtime
lrwxrwxrwx    1 root     root            25 Jul  9 20:09 /etc/localtime -> /opt/persistent/localtime
root@element-teone:~# ls -la /opt/persistent/localtime
lrwxrwxrwx    1 root     root            29 Aug  8 10:52 /opt/persistent/localtime -> /usr/share/zoneinfo/Universal
root@element-teone:~#
`

Reply:
By default the timezone is Universal.
```
root@AmlogicFirebolt:~# date
Fri Aug  9 04:21:27 UTC 2024
root@AmlogicFirebolt:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.getTimeZoneDST"} '
{"jsonrpc":"2.0","id":3,"result":{"timeZone":"Universal","accuracy":"INITIAL","success":true}}root@AmlogicFirebolt:~#
root@AmlogicFirebolt:~#
```


Tried to implement as per your suggestion by changing the softlink to corresponding timezone set:

```
if (!dirExists(dir)) {
        std::string command = "mkdir -p " + dir + " \0";
	command = "mkdir -p " + dir + " \0";
        Utils::cRunScript(command.c_str());
} else {
        //Do nothing//
	command = "ln -sf /usr/share/zoneinfo/" +timeZone +"  /etc/localtime" +" \0";
	Utils::cRunScript(command.c_str());
}                                
```
With the above change, tried to change timezone using org.rdk.System.setTimeZoneDST to Asia/Kolkate and then to America/New_York :
```
root@HP44HRDK:~# ls -lh /etc/localtime
lrwxrwxrwx    1 root     root          29 Aug  9 09:47 /etc/localtime -> /usr/share/zoneinfo/Universal
root@HP44HRDK:~#
root@HP44HRDK:~# date
Fri Aug  9 09:49:13 UTC 2024
root@HP44HRDK:~#
root@HP44HRDK:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.setTimeZoneDST", "params":{"timeZone": "Asia/Kolkata"}} '
{"jsonrpc":"2.0","id":3,"result":{"success":true}}root@HP44HRDK:~#
root@HP44HRDK:~#
root@HP44HRDK:~# ls -lh /etc/localtime
lrwxrwxrwx    1 root     root          32 Aug  9 15:19 /etc/localtime -> /usr/share/zoneinfo/Asia/Kolkata
root@HP44HRDK:~#
root@HP44HRDK:~# date
Fri Aug  9 15:19:43 IST 2024
root@HP44HRDK:~#
root@HP44HRDK:~# curl -X POST -H "Content-Type: application/json" 'http://127.0.0.1:9998/jsonrpc' -d ' {"jsonrpc":"2.0", "id":3, "method":"org.rdk.System.setTimeZoneDST", "params":{"timeZone": "America/New_York"}} '
{"jsonrpc":"2.0","id":3,"result":{"success":true}}root@HP44HRDK:~#
root@HP44HRDK:~#
root@HP44HRDK:~# date
Fri Aug  9 05:50:04 EDT 2024
root@HP44HRDK:~#
root@HP44HRDK:~# ls -lh /etc/localtime
lrwxrwxrwx    1 root     root          36 Aug  9 05:50 /etc/localtime -> /usr/share/zoneinfo/America/New_York
root@HP44HRDK:~#
```


Can you let me know if this is how you suggested @santoshcomcast?

Reply:
+1
Semms good.